agent/agent.go

```diff
@@ -225,6 +225,9 @@ func (a *Agent) run(ctx context.Context) {
 }
 
 func (a *Agent) handleSessionMessage(ctx context.Context, message *api.SessionMessage) error {
+	if message.Stop {
+		return a.Stop(context.Background())
```

Comment:
If two threads try to run that same code, only one can proceed to the `close` branch. The other will block and the `<-a.stopped` will proceed, avoiding the `close`.